File: harz-reach/v0.3-mesh/mesh-frame.js
// HARZ MESH FRAME v2.1 — byte-exact JS twin of the frozen Kotlin MeshFrame.kt
// (frozen G1/G2 v2.1 protocol, Sep 6 report; unchanged from commit 12b5991).
// Wire layout (all big-endian):
//   magic u32 0x48415A5A | type u8 | src 4B | dst 4B | seq u32 | ttl u8 | hops u8
//   routeLen u8 | route*4B | sigLen u8 (0|64) | sig | payLen u32 | payload | crc32 u32
// CRC32 over ALL preceding bytes. Cross-verified against the Kotlin core (InteropDriver).

import { crc32 } from "zlib";

export const MAGIC = 0x48415A5A;
export const PORT_GO = 8988;
export const CHUNK_SIZE = 900;
export const CHUNK_PACING_MS = 20;
export const DEFAULT_TTL = 8;
export const MAX_ROUTE = 16;
export const TYPE_HELLO = 0x01, TYPE_HELLO_ACK = 0x02, TYPE_DATA = 0x03, TYPE_ACK = 0x04, TYPE_BYE = 0x05;

function edgeBytes(id) {
  const s = String(id);
  if (s.length !== 4) throw new Error("EdgeId must be exactly 4 chars, got '" + s + "'");
  for (const ch of s) { const c = ch.charCodeAt(0); if (c < 0x20 || c > 0x7e) throw new Error("EdgeId must be printable ASCII"); }
  return Buffer.from(s, "ascii");
}
function edgeFrom(buf, off) {
  if (buf.length - off < 4) throw new Error("EdgeId needs 4 bytes");
  return buf.toString("ascii", off, off + 4);
}

export function encodeFrame(f) {
  const route = f.route || [];
  if (route.length > MAX_ROUTE) throw new Error("route overflow: " + route.length);
  const sig = f.sig || Buffer.alloc(0);
  if (sig.length !== 0 && sig.length !== 64) throw new Error("bad sigLen " + sig.length);
  const parts = [];
  const head = Buffer.alloc(17);
  head.writeUInt32BE(MAGIC, 0);
  head.writeUInt8(f.type, 4);
  edgeBytes(f.src).copy(head, 5);
  edgeBytes(f.dst).copy(head, 9);
  head.writeUInt32BE(f.seq >>> 0, 13);
  parts.push(head);
  const th = Buffer.alloc(2);
  th.writeUInt8(f.ttl, 0);
  th.writeUInt8(f.hops, 1);
  parts.push(th);
  parts.push(Buffer.from([route.length]));
  for (const e of route) parts.push(edgeBytes(e));
  parts.push(Buffer.from([sig.length]));
  parts.push(sig);
  const plen = Buffer.alloc(4);
  plen.writeUInt32BE(f.payload ? f.payload.length : 0, 0);
  parts.push(plen);
  parts.push(f.payload || Buffer.alloc(0));
  const body = Buffer.concat(parts);
  const crc = Buffer.alloc(4);
  crc.writeUInt32BE(crc32(body) >>> 0, 0);
  return Buffer.concat([body, crc]);
}

export function decodeFrame(raw) {
  if (raw.length < 29) throw new Error("frame too short: " + raw.length + "B");
  const magic = raw.readUInt32BE(0);
  if (magic !== MAGIC) throw new Error("bad magic 0x" + magic.toString(16).padStart(8, "0"));
  let off = 4;
  const type = raw.readUInt8(off); off += 1;
  const src = edgeFrom(raw, off); off += 4;
  const dst = edgeFrom(raw, off); off += 4;
  const seq = raw.readUInt32BE(off); off += 4;
  const ttl = raw.readUInt8(off); off += 1;
  const hops = raw.readUInt8(off); off += 1;
  const routeLen = raw.readUInt8(off); off += 1;
  if (routeLen > MAX_ROUTE) throw new Error("route overflow: " + routeLen);
  const route = [];
  for (let i = 0; i < routeLen; i++) { route.push(edgeFrom(raw, off)); off += 4; }
  const sigLen = raw.readUInt8(off); off += 1;
  if (sigLen !== 0 && sigLen !== 64) throw new Error("bad sigLen " + sigLen);
  const sig = raw.subarray(off, off + sigLen); off += sigLen;
  const payLen = raw.readUInt32BE(off); off += 4;
  if (payLen < 0 || payLen > 1_048_576) throw new Error("payload length out of bounds: " + payLen);
  const payload = raw.subarray(off, off + payLen); off += payLen;
  if (raw.length - off !== 4) throw new Error("trailing bytes after payload: " + (raw.length - off - 4));
  const crcStored = raw.readUInt32BE(off);
  const crcComputed = crc32(raw.subarray(0, off)) >>> 0;
  if (crcStored !== crcComputed) throw new Error("CRC mismatch: stored=" + crcStored + " computed=" + crcComputed);
  return { type, src, dst, seq, ttl, hops, route, sig, payload };
}

export const hello    = (id, seq = 0) => ({ type: TYPE_HELLO, src: id, dst: "    ", seq, ttl: 0, hops: 0, route: [] });
export const helloAck = (goId, to, seq) => ({ type: TYPE_HELLO_ACK, src: goId, dst: to, seq, ttl: 0, hops: 0, route: [] });
export const data     = (src, dst, seq, ttl, chunk) => ({ type: TYPE_DATA, src, dst, seq, ttl, hops: 0, route: [], payload: chunk });
export const ack      = (receiver, sender, seq, route) => ({ type: TYPE_ACK, src: receiver, dst: sender, seq, ttl: 0, hops: 0, route });
export const bye      = (id) => ({ type: TYPE_BYE, src: id, dst: "    ", seq: 0, ttl: 0, hops: 0, route: [] });

File: harz-edge-telecom/edge-rebuild-v2.2/core/src/main/kotlin/com/harz/edgetelecom/mesh/MeshFrame.kt
package com.harz.edgetelecom.mesh

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.util.zip.CRC32

/**
 * HARZ-Mesh wire frame.
 *
 * Layout (all big-endian):
 *   magic     u32   0x48415A5A
 *   type      u8    HELLO | HELLO_ACK | DATA | ACK | BYE
 *   src       4B    sender edge-ID
 *   dst       4B    destination edge-ID ("\0\0\0\0" = broadcast/local-any)
 *   seq       u32   per-sender chunk sequence number (ACKs echo it)
 *   ttl       u8    remaining hops-to-live (G2-C semantics in MeshCore)
 *   hops      u8    hop count, incremented by each forwarder
 *   routeLen  u8    number of 4-byte edge-IDs following
 *   route     routeLen * 4B   path of forwarders (excluding src)
 *   sigLen    u8    0 = unsigned (logic layer / G1-G2 runs), 64 = Ed25519 signature
 *   sig       sigLen bytes
 *   payLen    u32
 *   payload   payLen bytes
 *   crc32     u32   CRC32 over all preceding bytes
 *
 * The Sep 6 logic harness validated exactly these fields: dedup on (src,seq),
 * TTL rules, hop counting, route append, ACK return path, store-and-forward
 * of byte-identical payloads. Signatures exist in the frame but stay length-0
 * at the logic layer (IdentityManager activates them in the app build).
 */
data class MeshFrame(
    val type: Int,
    val src: EdgeId,
    val dst: EdgeId,
    val seq: Long,
    val ttl: Int,
    val hops: Int,
    val route: List<EdgeId>,
    val sig: ByteArray = ByteArray(0),
    val payload: ByteArray = ByteArray(0)
) {
    fun encode(): ByteArray {
        val buf = ByteBuffer.allocate(4 + 1 + 4 + 4 + 4 + 1 + 1 + 1 + route.size * 4 + 1 + sig.size + 4 + payload.size + 4)
        buf.putInt(MeshProtocol.MAGIC)
        buf.put(type.toByte())
        buf.put(src.bytes())
        buf.put(dst.bytes())
        buf.putInt(seq.toInt())
        buf.put(ttl.toByte())
        buf.put(hops.toByte())
        buf.put(route.size.toByte())
        for (e in route) buf.put(e.bytes())
        buf.put(sig.size.toByte())
        buf.put(sig)
        buf.putInt(payload.size)
        buf.put(payload)
        val body = buf.array().copyOfRange(0, buf.position())
        val crc = CRC32().apply { update(body) }
        return body + ByteBuffer.allocate(4).putInt(crc.value.toInt()).array()
    }

    override fun equals(other: Any?): Boolean = other is MeshFrame && other.encode().contentEquals(encode())
    override fun hashCode(): Int = encode().contentHashCode()

    companion object {
        /** Decode one frame from a complete byte buffer. Throws on any malformation. */
        fun decode(raw: ByteArray): MeshFrame {
            require(raw.size >= 29) { "frame too short: ${raw.size}B" }
            val bb = ByteBuffer.wrap(raw)
            val magic = bb.int
            require(magic == MeshProtocol.MAGIC) { "bad magic 0x%08x".format(magic) }
            val type = bb.get().toInt() and 0xFF
            val src = EdgeId.fromBytes(raw, bb.position()).also { if (true) bb.position(bb.position() + 4) }
            val dst = EdgeId.fromBytes(raw, bb.position()).also { if (true) bb.position(bb.position() + 4) }
            val seq = bb.int.toLong() and 0xFFFFFFFFL
            val ttl = bb.get().toInt() and 0xFF
            val hops = bb.get().toInt() and 0xFF
            val routeLen = bb.get().toInt() and 0xFF
            require(routeLen <= MeshProtocol.MAX_ROUTE) { "route overflow: $routeLen" }
            val route = ArrayList<EdgeId>(routeLen)
            repeat(routeLen) {
                route.add(EdgeId.fromBytes(raw, bb.position()))
                bb.position(bb.position() + 4)
            }
            val sigLen = bb.get().toInt() and 0xFF
            require(sigLen == 0 || sigLen == 64) { "bad sigLen $sigLen" }
            val sig = ByteArray(sigLen).also { bb.get(it) }
            val payLen = bb.int
            require(payLen in 0..1_048_576) { "payload length out of bounds: $payLen" }
            val payload = ByteArray(payLen).also { bb.get(it) }
            val crcOff = bb.position()
            require(raw.size - crcOff == 4) { "trailing bytes after payload: ${raw.size - crcOff - 4}" }
            val crcStored = bb.int.toLong() and 0xFFFFFFFFL
            val crc = CRC32().apply { update(raw, 0, crcOff) }
            require(crc.value == crcStored) { "CRC mismatch: stored=$crcStored computed=${crc.value}" }
            return MeshFrame(type, src, dst, seq, ttl, hops, route, sig, payload)
        }

        /** HELLO frame a new client sends to register its edge-ID. */
        fun hello(edgeId: EdgeId, seq: Long = 0): MeshFrame =
            MeshFrame(MeshProtocol.TYPE_HELLO, edgeId, EdgeId("    "), seq, 0, 0, emptyList())

        /** HELLO_ACK the GO answers with — carries the GO's own edge-ID. */
        fun helloAck(goEdgeId: EdgeId, to: EdgeId, seq: Long): MeshFrame =
            MeshFrame(MeshProtocol.TYPE_HELLO_ACK, goEdgeId, to, seq, 0, 0, emptyList())

        /** DATA chunk. */
        fun data(src: EdgeId, dst: EdgeId, seq: Long, ttl: Int, chunk: ByteArray): MeshFrame =
            MeshFrame(MeshProtocol.TYPE_DATA, src, dst, seq, ttl, 0, emptyList(), ByteArray(0), chunk)

        /** ACK — src=original receiver, dst=original sender, seq echoes the DATA seq. */
        fun ack(receiver: EdgeId, sender: EdgeId, seq: Long, route: List<EdgeId>): MeshFrame =
            MeshFrame(MeshProtocol.TYPE_ACK, receiver, sender, seq, 0, 0, route)

        fun bye(edgeId: EdgeId): MeshFrame =
            MeshFrame(MeshProtocol.TYPE_BYE, edgeId, EdgeId("    "), 0, 0, 0, emptyList())
    }
}
